Library/Control/Controller.cc

```diff
@@ -0,0 +1,154 @@
+/*
+ * Controller.cc
+ *
+ *  Created on: Jan 13, 2022
+ *      Author: N.S. Oblath
+ */
+
+#include "Controller.hh"
+
+#include "ControlAccess.hh"
+#include "Exception.hh"
+#include "QuitChain.hh"
+
+#include "logger.hh"
+
+
+namespace Nymph
+{
+    LOGGER( contlog, "Controller" );
+
+    Controller::Controller() :
+            ControllerInterface(),
+            scarab::cancelable(),
+            fCycleTimeMS( 500 ),
+            fMutex(),
+            fCondVarContinue(),
+            fCondVarBreak(),
+            fBreakFlag( false )
+    {
+        ControlAccess::get_instance()->SetControl( this );
+    }
+
+    Controller::~Controller()
+    {}
+
+    void Controller::Configure( const scarab::param_node& node )
+    {
+        SetCycleTimeMS( node.get_value( "cycle-time-ms", fCycleTimeMS ) );
+
+        return;
+    }
+
+    bool Controller::WaitToContinue()
+    {
+        std::unique_lock< std::mutex > lock( fMutex );
+        while( fBreakFlag && ! is_canceled() )
```

Comment:
Is there a structural reason that these to checks (break and canceled) are defined differently? There's nothing wrong with this approach, but is it reasonable to define `fCanceledFlag` instead for consistency?

Reply:
The `cancelable` base class is a part of Scarab, and it's really convenient to use it instead of a separately defining cancelation again.  Not only does it reduce duplicate code, it also makes it trivial to use with the existing `scarab::signal_handler`.  The important thing is that `Controller` now defines a uniform interface for checking for cancelation and breakpoints, so from the point of view of a client of `Controller`, everything looks consistent.

Reply:
Oh, I guess the think I left unsaid in the comment above is that the concept of a breakpoint, unlike cancelation, is specific to Nymph.  So while the concept of cancelation is pretty general and is used in a variety of situations in a more-or-less similar way, we're only going to see breakpoints (or at least this use of breakpoints) in Nymph.